src/storage/v2/replication/replication_client.cpp

```diff
@@ -343,37 +349,78 @@ void ReplicationStorageClient::Start(Storage *storage, DatabaseAccessProtector d
   TryCheckReplicaStateSync(storage, std::move(db_acc));
 }
 
-void ReplicationStorageClient::RecoverReplica(uint64_t replica_commit, memgraph::storage::Storage *storage) const {
+void ReplicationStorageClient::RecoverReplica(uint64_t replica_commit, Storage *storage) const {
   if (storage->storage_mode_ != StorageMode::IN_MEMORY_TRANSACTIONAL) {
     throw utils::BasicException("Only InMemoryTransactional mode supports replication!");
   }
   spdlog::debug("Starting replica {} recovery ", client_.name_);
   auto *mem_storage = static_cast<InMemoryStorage *>(storage);
 
+  // TODO: (andi) Don't like this while loop because it is blocking the queue
+  rpc::Client &rpcClient = client_.rpc_client_;
+  bool recovery_failed{false};
   while (true) {
+    if (recovery_failed) {
+      return;
+    }
     auto file_locker = mem_storage->file_retainer_.AddLocker();
 
     const auto steps = GetRecoverySteps(replica_commit, &file_locker, mem_storage);
-    int i = 0;
-    for (const RecoveryStep &recovery_step : steps) {
-      spdlog::trace("Recovering in step: {}", i++);
+    for (auto const &[step_index, recovery_step] : ranges::views::enumerate(steps)) {
+      // If recovery failed, set the state to MAYBE_BEHIND because replica for sure didn't recovery completely
+      if (recovery_failed) {
+        spdlog::debug("One of recovery steps failed, setting replica state to MAYBE_BEHIND");
+        replica_state_.WithLock([](auto &val) { val = ReplicaState::MAYBE_BEHIND; });
+        return;
+      }
       try {
-        rpc::Client &rpcClient = client_.rpc_client_;
+        spdlog::trace("Recovering in step: {}", step_index);
         std::visit(
             utils::Overloaded{
-                [this, &replica_commit, mem_storage, &rpcClient,
+                [this, &replica_commit, mem_storage, &rpcClient, &recovery_failed,
                  main_uuid = main_uuid_](RecoverySnapshot const &snapshot) {
                   spdlog::debug("Sending the latest snapshot file: {} to {}", snapshot, client_.name_);
-                  auto response = TransferSnapshot(main_uuid, mem_storage->uuid(), rpcClient, snapshot);
-                  replica_commit = response.current_commit_timestamp;
```

Comment:
If loading snapshot passes, all is OK and same as before. If loading snapshot fails, before replica_commit was set to 0, now it's left as it was. I think this is ok.